File: Java/net/zedmeelab/zedmeehash/ZedmeeHash64.java
package net.zedmeelab.zedmeehash;

/**
 * ZedmeeHash64
 * Very strong, fast, non-cryptographic 64-bit hash function
 *
 * Copyright(C) 2021 - Matteo Zapparoli
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */
public class ZedmeeHash64 extends ZedmeeHash {

	public static final int DEFAULT_SEED = 1473841489;

	public ZedmeeHash64(int seed1, int seed2, int seed3, int seed4) {
		super(seed1, seed2, seed3, seed4);
	}

	public ZedmeeHash64(int seed) {
		super(seed);
	}

	public ZedmeeHash64() {
		super(DEFAULT_SEED);
	}

	/**
	 * Evaluate and return the hash value of [data[pos], data[pos+1] ... data[pos + length - 1]]
	 * @param data
	 * @param pos >= 0 && pos < data.length
	 * @param length >= 0 && pos + length <= data.length
	 * @return hash value
	 */
	public long hash(final byte[] data, int pos, int length) {
		long h = 1L;
		
		while(length > 0) {
    		long b1, b2, b3, b4, b5, b6, b7, b8;
    
    		switch(length) {
    		case 1:
    			b1 = b2 = b3 = b4 = b5 = b6 = b7 = byteTable0;
    			b8 = byteTable[data[pos] & 0xff];
    			length = 0;
    			break;
    
    		case 2:
    			b1 = b2 = b3 = b4 = b5 = b6 = byteTable0;
    			b7 = byteTable[data[pos] & 0xff];
    			b8 = byteTable[data[pos + 1] & 0xff];
    			length = 0;
    			break;
    
    		case 3:
    			b1 = b2 = b3 = b4 = b5 = byteTable0;
    			b6 = byteTable[data[pos] & 0xff];
    			b7 = byteTable[data[pos + 1] & 0xff];
    			b8 = byteTable[data[pos + 2] & 0xff];
    			length = 0;
    			break;
    
    		case 4:
    			b1 = b2 = b3 = b4 = byteTable0;
    			b5 = byteTable[data[pos] & 0xff];
    			b6 = byteTable[data[pos + 1] & 0xff];
    			b7 = byteTable[data[pos + 2] & 0xff];
    			b8 = byteTable[data[pos + 3] & 0xff];
    			length = 0;
    			break;
    
    		case 5:
    			b1 = b2 = b3 = byteTable0;
    			b4 = byteTable[data[pos] & 0xff];
    			b5 = byteTable[data[pos + 1] & 0xff];
    			b6 = byteTable[data[pos + 2] & 0xff];
    			b7 = byteTable[data[pos + 3] & 0xff];
    			b8 = byteTable[data[pos + 4] & 0xff];
    			length = 0;
    			break;
    
    		case 6:
    			b1 = b2 = byteTable0;
    			b3 = byteTable[data[pos] & 0xff];
    			b4 = byteTable[data[pos + 1] & 0xff];
    			b5 = byteTable[data[pos + 2] & 0xff];
    			b6 = byteTable[data[pos + 3] & 0xff];
    			b7 = byteTable[data[pos + 4] & 0xff];
    			b8 = byteTable[data[pos + 5] & 0xff];
    			length = 0;
    			break;
    
    		case 7:
    			b1 = byteTable0;
    			b2 = byteTable[data[pos] & 0xff];
    			b3 = byteTable[data[pos + 1] & 0xff];
    			b4 = byteTable[data[pos + 2] & 0xff];
    			b5 = byteTable[data[pos + 3] & 0xff];
    			b6 = byteTable[data[pos + 4] & 0xff];
    			b7 = byteTable[data[pos + 5] & 0xff];
    			b8 = byteTable[data[pos + 6] & 0xff];
    			length = 0;
    			break;
    
    		default:
    			b1 = byteTable[(pos + data[pos++]) & 0xff];
    			b2 = byteTable[(pos + data[pos++]) & 0xff];
    			b3 = byteTable[(pos + data[pos++]) & 0xff];
    			b4 = byteTable[(pos + data[pos++]) & 0xff];
    			b5 = byteTable[(pos + data[pos++]) & 0xff];
    			b6 = byteTable[(pos + data[pos++]) & 0xff];
    			b7 = byteTable[(pos + data[pos++]) & 0xff];
    			b8 = byteTable[(pos + data[pos++]) & 0xff];
    			length -= 8;
    		}
    
    		long x1 = b2 ^ b1;
    		long x2 = b4 ^ b3;
    		long x3 = b6 ^ b5;
    		long x4 = b8 ^ b7;
    		long y1 = x2 ^ x1;
    		long y2 = x4 ^ x3;
    
    		h = (6364136223846793005L * h) ^
    		    (((b7 ^ x3 ^ y1) << 56) |
    			 ((b8 ^ x3 ^ y1) << 48) |
    			 ((x4 ^ b5 ^ y1) << 40) |
    			 ((x4 ^ b6 ^ y1) << 32) |
    			 ((y2 ^ b3 ^ x1) << 24) |
    			 ((y2 ^ b4 ^ x1) << 16) |
    			 ((y2 ^ x2 ^ b1) << 8) |
    			 ((y2 ^ x2 ^ b2)));
		}

		return h;
	}

	public long hash(final byte[] data, int length) {
		return hash(data, 0, length);
	}

	public long hash(final byte[] data) {
		return hash(data, 0, data.length);
	}

}

File: zedmeehash64.c
/**
 * ZedmeeHash64 
 * Strong, fast, simple, non-cryptographic 32-bit hash function
 * 
 * Author: Matteo Zapparoli
 * Date: 2021
 * Licence: Public Domain
 * 
 * This is free and unencumbered software released into the public domain.
 *
 * Anyone is free to copy, modify, publish, use, compile, sell, or
 * distribute this software, either in source code form or as a compiled
 * binary, for any purpose, commercial or non-commercial, and by any
 * means.
 * 
 * In jurisdictions that recognize copyright laws, the author or authors
 * of this software dedicate any and all copyright interest in the
 * software to the public domain. We make this dedication for the benefit
 * of the public at large and to the detriment of our heirs and
 * successors. We intend this dedication to be an overt act of
 * relinquishment in perpetuity of all present and future rights to this
 * software under copyright law.
 * 
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
 * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
 * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
 * IN NO EVENT SHALL THE AUTHORS BE LIABLE FOR ANY CLAIM, DAMAGES OR
 * OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
 * ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
 * OTHER DEALINGS IN THE SOFTWARE.
 * 
 * For more information, please refer to <https://unlicense.org>
 * 
 */
 
#include <stdlib.h> // size_t
#include <stdint.h> // uint64_t

#include "zedmeehash64.h"
 
uint64_t default_table64[256];
 
/**
 * Generate the lookup table of 256 uint64_t using the lfsr258 algorithm
 * param table: uint64_t[256]
 **/
void zmh64create_table(uint64_t table[], uint64_t seed1, uint64_t seed2, 
                       uint64_t seed3, uint64_t seed4, uint64_t seed5)
{
	// lfsr258 pseudo random number generator
	// Author: Pierre L'Ecuyer
	// The initial seeds y1, y2, y3, y4, y5  MUST be larger than
	// 1, 511, 4095, 131071 and 8388607 respectively.
	
	// if seed1 is less than 2 
	if((seed1 & 0xFFFFFFFFFFFFFFFEL) == 0) seed1 |= 0x02L;
	
	// if seed2 is less than 512 
	if((seed2 & 0xFFFFFFFFFFFFFE00L) == 0) seed2 |= 0x0200L;
	
	// if seed3 is less than 4096 
	if((seed3 & 0xFFFFFFFFFFFFF000L) == 0) seed3 |= 0x01000L;
	
	// if seed4 is less than 131072
	if((seed4 & 0xFFFFFFFFFFFE0000L) == 0) seed4 |= 0x20000L;
	
	// if seed5 is less than 8388608
	if((seed5 & 0xFFFFFFFFFF800000L) == 0) seed5 |= 0x800000L;
	
	for(int i = 0; i < 256; i++) 
	{
		uint64_t b = ((seed1 << 1) ^ seed1) >> 53;
		seed1 = ((seed1 & 0xFFFFFFFFFFFFFFFEL) << 10) ^ b;
		b = ((seed2 << 24) ^ seed2) >> 50;
		seed2 = ((seed2 & 0xFFFFFFFFFFFFFE00L) << 5) ^ b;
		b = ((seed3 << 3) ^ seed3) >> 23;
		seed3 = ((seed3 & 0xFFFFFFFFFFFFF000L) << 29) ^ b;
		b = ((seed4 << 5) ^ seed4) >> 24;
		seed4 = ((seed4 & 0xFFFFFFFFFFFE0000L) << 23) ^ b;
		b = ((seed5 << 3) ^ seed5) >> 33;
		seed5 = ((seed5 & 0xFFFFFFFFFF800000L) << 8) ^ b;
		
		table[i] = (seed1 ^ seed2 ^ seed3 ^ seed4 ^ seed5);
	}
}							
 
/**
 * Initialize the defalt table
 */
void zmh64init_table(void)
{
	zmh64create_table(default_table64, DEFAULT_TABLE_SEED64_1, DEFAULT_TABLE_SEED64_2,
	                 DEFAULT_TABLE_SEED64_3, DEFAULT_TABLE_SEED64_4, DEFAULT_TABLE_SEED64_5);
}

